File: adk/node/tests/audit.test.ts
/**
 * ZAK Audit Events + Logger Tests
 *
 * Covers:
 * - Event factory functions create correct event types
 * - AuditEvent base fields are populated
 * - AuditLogger emits events without crashing
 * - AuditLogger logRaw works
 * - AuditEventType constants
 */

import { describe, it, expect } from "vitest";

import {
  AuditEventType,
  agentStartedEvent,
  agentCompletedEvent,
  agentFailedEvent,
  policyBlockedEvent,
  toolCalledEvent,
  graphWriteEvent,
} from "../src/core/audit/events.js";
import type {
  AuditEvent,
  AgentStartedEvent,
  AgentCompletedEvent,
  AgentFailedEvent,
  PolicyBlockedEvent,
  ToolCalledEvent,
  GraphWriteEvent,
} from "../src/core/audit/events.js";
import { AuditLogger } from "../src/core/audit/logger.js";

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------
const AGENT_ID = "test-agent-v1";
const TENANT_ID = "acme-corp";
const TRACE_ID = "trace-abc-123";

// ---------------------------------------------------------------------------
// AuditEventType constants
// ---------------------------------------------------------------------------
describe("AuditEventType constants", () => {
  it("has AGENT_STARTED type", () => {
    expect(AuditEventType.AGENT_STARTED).toBe("agent.started");
  });

  it("has AGENT_COMPLETED type", () => {
    expect(AuditEventType.AGENT_COMPLETED).toBe("agent.completed");
  });

  it("has AGENT_FAILED type", () => {
    expect(AuditEventType.AGENT_FAILED).toBe("agent.failed");
  });

  it("has TOOL_CALLED type", () => {
    expect(AuditEventType.TOOL_CALLED).toBe("agent.tool_called");
  });

  it("has TOOL_RESULT type", () => {
    expect(AuditEventType.TOOL_RESULT).toBe("agent.tool_result");
  });

  it("has DECISION_MADE type", () => {
    expect(AuditEventType.DECISION_MADE).toBe("agent.decision");
  });

  it("has POLICY_ALLOWED type", () => {
    expect(AuditEventType.POLICY_ALLOWED).toBe("policy.allowed");
  });

  it("has POLICY_BLOCKED type", () => {
    expect(AuditEventType.POLICY_BLOCKED).toBe("policy.blocked");
  });

  it("has GRAPH_READ type", () => {
    expect(AuditEventType.GRAPH_READ).toBe("sif.graph_read");
  });

  it("has GRAPH_WRITE type", () => {
    expect(AuditEventType.GRAPH_WRITE).toBe("sif.graph_write");
  });

  it("has HUMAN_APPROVAL_REQUESTED type", () => {
    expect(AuditEventType.HUMAN_APPROVAL_REQUESTED).toBe(
      "governance.approval_requested"
    );
  });

  it("has HUMAN_APPROVAL_GRANTED type", () => {
    expect(AuditEventType.HUMAN_APPROVAL_GRANTED).toBe(
      "governance.approval_granted"
    );
  });

  it("has HUMAN_APPROVAL_DENIED type", () => {
    expect(AuditEventType.HUMAN_APPROVAL_DENIED).toBe(
      "governance.approval_denied"
    );
  });

  it("has exactly 13 event types", () => {
    const keys = Object.keys(AuditEventType);
    expect(keys.length).toBe(13);
  });
});

// ---------------------------------------------------------------------------
// Event factory: agentStartedEvent
// ---------------------------------------------------------------------------
describe("agentStartedEvent factory", () => {
  it("creates an event with correct eventType", () => {
    const event = agentStartedEvent(AGENT_ID, TENANT_ID, TRACE_ID);
    expect(event.eventType).toBe(AuditEventType.AGENT_STARTED);
  });

  it("populates agentId, tenantId, traceId", () => {
    const event = agentStartedEvent(AGENT_ID, TENANT_ID, TRACE_ID);
    expect(event.agentId).toBe(AGENT_ID);
    expect(event.tenantId).toBe(TENANT_ID);
    expect(event.traceId).toBe(TRACE_ID);
  });

  it("generates an ISO 8601 timestamp", () => {
    const event = agentStartedEvent(AGENT_ID, TENANT_ID, TRACE_ID);
    expect(event.timestamp).toBeTruthy();
    // Should be parseable as a date
    const date = new Date(event.timestamp);
    expect(date.getTime()).not.toBeNaN();
  });

  it("has an empty payload by default", () => {
    const event = agentStartedEvent(AGENT_ID, TENANT_ID, TRACE_ID);
    expect(event.payload).toEqual({});
  });

  it("accepts custom payload", () => {
    const event = agentStartedEvent(AGENT_ID, TENANT_ID, TRACE_ID, {
      trigger: "manual",
    });
    expect(event.payload).toEqual({ trigger: "manual" });
  });
});

// ---------------------------------------------------------------------------
// Event factory: agentCompletedEvent
// ---------------------------------------------------------------------------
describe("agentCompletedEvent factory", () => {
  it("creates an event with correct eventType", () => {
    const event = agentCompletedEvent(AGENT_ID, TENANT_ID, TRACE_ID, true, 1500);
    expect(event.eventType).toBe(AuditEventType.AGENT_COMPLETED);
  });

  it("includes success and durationMs fields", () => {
    const event = agentCompletedEvent(AGENT_ID, TENANT_ID, TRACE_ID, true, 1500);
    expect(event.success).toBe(true);
    expect(event.durationMs).toBe(1500);
  });

  it("records failure with success=false", () => {
    const event = agentCompletedEvent(AGENT_ID, TENANT_ID, TRACE_ID, false, 500);
    expect(event.success).toBe(false);
    expect(event.durationMs).toBe(500);
  });

  it("populates base fields correctly", () => {
    const event = agentCompletedEvent(AGENT_ID, TENANT_ID, TRACE_ID, true, 100);
    expect(event.agentId).toBe(AGENT_ID);
    expect(event.tenantId).toBe(TENANT_ID);
    expect(event.traceId).toBe(TRACE_ID);
    expect(event.timestamp).toBeTruthy();
  });

  it("accepts custom payload", () => {
    const event = agentCompletedEvent(AGENT_ID, TENANT_ID, TRACE_ID, true, 200, {
      findings: 5,
    });
    expect(event.payload).toEqual({ findings: 5 });
  });
});

// ---------------------------------------------------------------------------
// Event factory: agentFailedEvent
// ---------------------------------------------------------------------------
describe("agentFailedEvent factory", () => {
  it("creates an event with correct eventType", () => {
    const event = agentFailedEvent(AGENT_ID, TENANT_ID, TRACE_ID, "Connection timeout");
    expect(event.eventType).toBe(AuditEventType.AGENT_FAILED);
  });

  it("includes the error message", () => {
    const event = agentFailedEvent(AGENT_ID, TENANT_ID, TRACE_ID, "Out of memory");
    expect(event.error).toBe("Out of memory");
  });

  it("populates base fields correctly", () => {
    const event = agentFailedEvent(AGENT_ID, TENANT_ID, TRACE_ID, "Error");
    expect(event.agentId).toBe(AGENT_ID);
    expect(event.tenantId).toBe(TENANT_ID);
    expect(event.traceId).toBe(TRACE_ID);
  });

  it("accepts custom payload", () => {
    const event = agentFailedEvent(AGENT_ID, TENANT_ID, TRACE_ID, "Error", {
      retryCount: 3,
    });
    expect(event.payload).toEqual({ retryCount: 3 });
  });
});

// ---------------------------------------------------------------------------
// Event factory: policyBlockedEvent
// ---------------------------------------------------------------------------
describe("policyBlockedEvent factory", () => {
  it("creates an event with correct eventType", () => {
    const event = policyBlockedEvent(
      AGENT_ID,
      TENANT_ID,
      TRACE_ID,
      "execute_exploit",
      "Risk budget too low"
    );
    expect(event.eventType).toBe(AuditEventType.POLICY_BLOCKED);
  });

  it("includes action and reason fields", () => {
    const event = policyBlockedEvent(
      AGENT_ID,
      TENANT_ID,
      TRACE_ID,
      "deploy_payload",
      "Denied by policy"
    );
    expect(event.action).toBe("deploy_payload");
    expect(event.reason).toBe("Denied by policy");
  });

  it("populates base fields correctly", () => {
    const event = policyBlockedEvent(
      AGENT_ID,
      TENANT_ID,
      TRACE_ID,
      "action",
      "reason"
    );
    expect(event.agentId).toBe(AGENT_ID);
    expect(event.tenantId).toBe(TENANT_ID);
    expect(event.traceId).toBe(TRACE_ID);
  });

  it("accepts custom payload", () => {
    const event = policyBlockedEvent(
      AGENT_ID,
      TENANT_ID,
      TRACE_ID,
      "action",
      "reason",
      { environment: "production" }
    );
    expect(event.payload).toEqual({ environment: "production" });
  });
});

// ---------------------------------------------------------------------------
// Event factory: toolCalledEvent
// ---------------------------------------------------------------------------
describe("toolCalledEvent factory", () => {
  it("creates an event with correct eventType", () => {
    const event = toolCalledEvent(
      AGENT_ID,
      TENANT_ID,
      TRACE_ID,
      "nmap_scan",
      "Scanning target 192.168.1.0/24"
    );
    expect(event.eventType).toBe(AuditEventType.TOOL_CALLED);
  });

  it("includes toolName and inputSummary fields", () => {
    const event = toolCalledEvent(
      AGENT_ID,
      TENANT_ID,
      TRACE_ID,
      "risk_calculator",
      "Computing ALE for scenario S-001"
    );
    expect(event.toolName).toBe("risk_calculator");
    expect(event.inputSummary).toBe("Computing ALE for scenario S-001");
  });

  it("populates base fields correctly", () => {
    const event = toolCalledEvent(
      AGENT_ID,
      TENANT_ID,
      TRACE_ID,
      "tool",
      "summary"
    );
    expect(event.agentId).toBe(AGENT_ID);
    expect(event.tenantId).toBe(TENANT_ID);
    expect(event.traceId).toBe(TRACE_ID);
  });
});

// ---------------------------------------------------------------------------
// Event factory: graphWriteEvent
// ---------------------------------------------------------------------------
describe("graphWriteEvent factory", () => {
  it("creates an event with correct eventType", () => {
    const event = graphWriteEvent(
      AGENT_ID,
      TENANT_ID,
      TRACE_ID,
      "Vulnerability",
      "vuln-001"
    );
    expect(event.eventType).toBe(AuditEventType.GRAPH_WRITE);
  });

  it("includes nodeType, nodeId, and operation fields", () => {
    const event = graphWriteEvent(
      AGENT_ID,
      TENANT_ID,
      TRACE_ID,
      "Asset",
      "asset-42",
      "create"
    );
    expect(event.nodeType).toBe("Asset");
    expect(event.nodeId).toBe("asset-42");
    expect(event.operation).toBe("create");
  });

  it("defaults operation to upsert", () => {
    const event = graphWriteEvent(
      AGENT_ID,
      TENANT_ID,
      TRACE_ID,
      "Risk",
      "risk-001"
    );
    expect(event.operation).toBe("upsert");
  });

  it("populates base fields correctly", () => {
    const event = graphWriteEvent(
      AGENT_ID,
      TENANT_ID,
      TRACE_ID,
      "Node",
      "id-1"
    );
    expect(event.agentId).toBe(AGENT_ID);
    expect(event.tenantId).toBe(TENANT_ID);
    expect(event.traceId).toBe(TRACE_ID);
  });

  it("accepts custom payload", () => {
    const event = graphWriteEvent(
      AGENT_ID,
      TENANT_ID,
      TRACE_ID,
      "Asset",
      "asset-1",
      "upsert",
      { properties: { name: "web-server" } }
    );
    expect(event.payload).toEqual({ properties: { name: "web-server" } });
  });
});

// ---------------------------------------------------------------------------
// AuditLogger — emit events without crashing
// ---------------------------------------------------------------------------
describe("AuditLogger", () => {
  it("constructs with tenantId, agentId, and traceId", () => {
    const logger = new AuditLogger(TENANT_ID, AGENT_ID, TRACE_ID);
    expect(logger.tenantId).toBe(TENANT_ID);
    expect(logger.agentId).toBe(AGENT_ID);
    expect(logger.traceId).toBe(TRACE_ID);
  });

  it("emits agentStartedEvent without throwing", () => {
    const logger = new AuditLogger(TENANT_ID, AGENT_ID, TRACE_ID);
    const event = agentStartedEvent(AGENT_ID, TENANT_ID, TRACE_ID);
    expect(() => logger.emit(event)).not.toThrow();
  });

  it("emits agentCompletedEvent without throwing", () => {
    const logger = new AuditLogger(TENANT_ID, AGENT_ID, TRACE_ID);
    const event = agentCompletedEvent(AGENT_ID, TENANT_ID, TRACE_ID, true, 1000);
    expect(() => logger.emit(event)).not.toThrow();
  });

  it("emits agentFailedEvent without throwing", () => {
    const logger = new AuditLogger(TENANT_ID, AGENT_ID, TRACE_ID);
    const event = agentFailedEvent(AGENT_ID, TENANT_ID, TRACE_ID, "Test error");
    expect(() => logger.emit(event)).not.toThrow();
  });

  it("emits policyBlockedEvent without throwing", () => {
    const logger = new AuditLogger(TENANT_ID, AGENT_ID, TRACE_ID);
    const event = policyBlockedEvent(
      AGENT_ID,
      TENANT_ID,
      TRACE_ID,
      "execute_exploit",
      "Denied by policy"
    );
    expect(() => logger.emit(event)).not.toThrow();
  });

  it("emits toolCalledEvent without throwing", () => {
    const logger = new AuditLogger(TENANT_ID, AGENT_ID, TRACE_ID);
    const event = toolCalledEvent(
      AGENT_ID,
      TENANT_ID,
      TRACE_ID,
      "nmap",
      "scan target"
    );
    expect(() => logger.emit(event)).not.toThrow();
  });

  it("emits graphWriteEvent without throwing", () => {
    const logger = new AuditLogger(TENANT_ID, AGENT_ID, TRACE_ID);
    const event = graphWriteEvent(
      AGENT_ID,
      TENANT_ID,
      TRACE_ID,
      "Vulnerability",
      "vuln-001"
    );
    expect(() => logger.emit(event)).not.toThrow();
  });

  it("logRaw emits without throwing", () => {
    const logger = new AuditLogger(TENANT_ID, AGENT_ID, TRACE_ID);
    expect(() =>
      logger.logRaw(AuditEventType.AGENT_STARTED, { custom: "data" })
    ).not.toThrow();
  });

  it("logRaw works with empty data", () => {
    const logger = new AuditLogger(TENANT_ID, AGENT_ID, TRACE_ID);
    expect(() =>
      logger.logRaw(AuditEventType.DECISION_MADE)
    ).not.toThrow();
  });

  it("emits multiple events sequentially without throwing", () => {
    const logger = new AuditLogger(TENANT_ID, AGENT_ID, TRACE_ID);
    const started = agentStartedEvent(AGENT_ID, TENANT_ID, TRACE_ID);
    const tool = toolCalledEvent(AGENT_ID, TENANT_ID, TRACE_ID, "scanner", "scan");
    const completed = agentCompletedEvent(AGENT_ID, TENANT_ID, TRACE_ID, true, 2000);

    expect(() => {
      logger.emit(started);
      logger.emit(tool);
      logger.emit(completed);
    }).not.toThrow();
  });
});

// ---------------------------------------------------------------------------
// Timestamp consistency
// ---------------------------------------------------------------------------
describe("Timestamp consistency", () => {
  it("generates timestamps close to current time", () => {
    const before = Date.now();
    const event = agentStartedEvent(AGENT_ID, TENANT_ID, TRACE_ID);
    const after = Date.now();
    const eventTime = new Date(event.timestamp).getTime();
    expect(eventTime).toBeGreaterThanOrEqual(before);
    expect(eventTime).toBeLessThanOrEqual(after);
  });

  it("different events get different timestamps (or same if called fast)", () => {
    const event1 = agentStartedEvent(AGENT_ID, TENANT_ID, TRACE_ID);
    const event2 = agentStartedEvent(AGENT_ID, TENANT_ID, TRACE_ID);
    // Both should be valid ISO strings
    expect(new Date(event1.timestamp).getTime()).not.toBeNaN();
    expect(new Date(event2.timestamp).getTime()).not.toBeNaN();
  });
});

File: adk/go/pkg/audit/audit_test.go
package audit

import (
	"bytes"
	"encoding/json"
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

// ---------------------------------------------------------------------------
// Event factory functions — correct types and fields
// ---------------------------------------------------------------------------

func TestNewEvent_CreatesCorrectType(t *testing.T) {
	e := NewEvent(AgentStarted, "agent-1", "tenant-1", "trace-1")

	assert.Equal(t, AgentStarted, e.EventType)
	assert.Equal(t, "agent-1", e.AgentID)
	assert.Equal(t, "tenant-1", e.TenantID)
	assert.Equal(t, "trace-1", e.TraceID)
	assert.NotNil(t, e.Payload)
	assert.WithinDuration(t, time.Now().UTC(), e.Timestamp, 2*time.Second)
}

func TestNewEvent_WithPayload(t *testing.T) {
	e := NewEvent(ToolCalled, "agent-1", "tenant-1", "trace-1").
		WithPayload("tool_name", "nmap").
		WithPayload("status", "running")

	assert.Equal(t, "nmap", e.Payload["tool_name"])
	assert.Equal(t, "running", e.Payload["status"])
}

func TestAgentStartedEvent(t *testing.T) {
	e := AgentStartedEvent("agent-1", "tenant-1", "trace-1")
	assert.Equal(t, AgentStarted, e.EventType)
	assert.Equal(t, "agent-1", e.AgentID)
	assert.Equal(t, "tenant-1", e.TenantID)
	assert.Equal(t, "trace-1", e.TraceID)
}

func TestAgentCompletedEvent(t *testing.T) {
	e := AgentCompletedEvent("agent-1", "tenant-1", "trace-1", true, 1234.5)
	assert.Equal(t, AgentCompleted, e.EventType)
	assert.Equal(t, true, e.Payload["success"])
	assert.Equal(t, 1234.5, e.Payload["duration_ms"])
}

func TestAgentCompletedEvent_Failure(t *testing.T) {
	e := AgentCompletedEvent("agent-1", "tenant-1", "trace-1", false, 500.0)
	assert.Equal(t, AgentCompleted, e.EventType)
	assert.Equal(t, false, e.Payload["success"])
	assert.Equal(t, 500.0, e.Payload["duration_ms"])
}

func TestAgentFailedEvent(t *testing.T) {
	e := AgentFailedEvent("agent-1", "tenant-1", "trace-1", "timeout exceeded")
	assert.Equal(t, AgentFailed, e.EventType)
	assert.Equal(t, "timeout exceeded", e.Payload["error"])
}

func TestPolicyBlockedEvent(t *testing.T) {
	e := PolicyBlockedEvent("agent-1", "tenant-1", "trace-1", "deploy_payload", "denied by policy")
	assert.Equal(t, PolicyBlocked, e.EventType)
	assert.Equal(t, "deploy_payload", e.Payload["action"])
	assert.Equal(t, "denied by policy", e.Payload["reason"])
}

func TestToolCalledEvent(t *testing.T) {
	e := ToolCalledEvent("agent-1", "tenant-1", "trace-1", "nmap_scan", "target=10.0.0.1")
	assert.Equal(t, ToolCalled, e.EventType)
	assert.Equal(t, "nmap_scan", e.Payload["tool_name"])
	assert.Equal(t, "target=10.0.0.1", e.Payload["input_summary"])
}

func TestGraphWriteEvent(t *testing.T) {
	e := GraphWriteEvent("agent-1", "tenant-1", "trace-1", "Vulnerability", "vuln-123", "create")
	assert.Equal(t, GraphWrite, e.EventType)
	assert.Equal(t, "Vulnerability", e.Payload["node_type"])
	assert.Equal(t, "vuln-123", e.Payload["node_id"])
	assert.Equal(t, "create", e.Payload["operation"])
}

// ---------------------------------------------------------------------------
// EventType constants
// ---------------------------------------------------------------------------

func TestEventTypeConstants(t *testing.T) {
	expectedTypes := map[EventType]string{
		AgentStarted:           "agent.started",
		AgentCompleted:         "agent.completed",
		AgentFailed:            "agent.failed",
		ToolCalled:             "agent.tool_called",
		ToolResult:             "agent.tool_result",
		DecisionMade:           "agent.decision",
		PolicyAllowed:          "policy.allowed",
		PolicyBlocked:          "policy.blocked",
		GraphRead:              "sif.graph_read",
		GraphWrite:             "sif.graph_write",
		HumanApprovalRequested: "governance.approval_requested",
		HumanApprovalGranted:   "governance.approval_granted",
		HumanApprovalDenied:    "governance.approval_denied",
	}

	for et, expected := range expectedTypes {
		assert.Equal(t, expected, string(et), "EventType %q should have value %q", et, expected)
	}
}

// ---------------------------------------------------------------------------
// Logger — Emit events
// ---------------------------------------------------------------------------

func TestLogger_EmitWritesJSON(t *testing.T) {
	var buf bytes.Buffer
	logger := NewLoggerWithWriter("acme", "risk-quant-v1", "trace-abc", &buf)

	event := AgentStartedEvent("risk-quant-v1", "acme", "trace-abc")
	logger.Emit(event)

	output := buf.String()
	require.NotEmpty(t, output, "logger should write output")

	// Parse the JSON output
	var logLine map[string]interface{}
	err := json.Unmarshal([]byte(output), &logLine)
	require.NoError(t, err, "output should be valid JSON")

	assert.Equal(t, "agent.started", logLine["message"])
	assert.Equal(t, "acme", logLine["tenant_id"])
	assert.Equal(t, "risk-quant-v1", logLine["agent_id"])
	assert.Equal(t, "trace-abc", logLine["trace_id"])
	assert.Contains(t, logLine, "timestamp")
}

func TestLogger_EmitIncludesPayload(t *testing.T) {
	var buf bytes.Buffer
	logger := NewLoggerWithWriter("acme", "agent-1", "trace-1", &buf)

	event := AgentCompletedEvent("agent-1", "acme", "trace-1", true, 999.0)
	logger.Emit(event)

	var logLine map[string]interface{}
	err := json.Unmarshal(buf.Bytes(), &logLine)
	require.NoError(t, err)

	assert.Equal(t, "agent.completed", logLine["message"])
	assert.Equal(t, true, logLine["success"])
	assert.Equal(t, 999.0, logLine["duration_ms"])
}

func TestLogger_EmitMultipleEvents(t *testing.T) {
	var buf bytes.Buffer
	logger := NewLoggerWithWriter("acme", "agent-1", "trace-1", &buf)

	logger.Emit(AgentStartedEvent("agent-1", "acme", "trace-1"))
	logger.Emit(ToolCalledEvent("agent-1", "acme", "trace-1", "scanner", "target=x"))
	logger.Emit(AgentCompletedEvent("agent-1", "acme", "trace-1", true, 1000.0))

	lines := bytes.Split(bytes.TrimSpace(buf.Bytes()), []byte("\n"))
	assert.Len(t, lines, 3, "should emit exactly 3 log lines")

	// Verify each line is valid JSON
	for i, line := range lines {
		var logLine map[string]interface{}
		err := json.Unmarshal(line, &logLine)
		assert.NoError(t, err, "line %d should be valid JSON", i)
	}
}

// ---------------------------------------------------------------------------
// Logger — LogRaw
// ---------------------------------------------------------------------------

func TestLogger_LogRaw(t *testing.T) {
	var buf bytes.Buffer
	logger := NewLoggerWithWriter("acme", "agent-1", "trace-1", &buf)

	fields := map[string]interface{}{
		"custom_field": "custom_value",
		"count":        42,
	}
	logger.LogRaw(DecisionMade, fields)

	output := buf.String()
	require.NotEmpty(t, output)

	var logLine map[string]interface{}
	err := json.Unmarshal([]byte(output), &logLine)
	require.NoError(t, err, "output should be valid JSON")

	assert.Equal(t, "agent.decision", logLine["message"])
	assert.Equal(t, "custom_value", logLine["custom_field"])
	// JSON numbers are parsed as float64
	assert.Equal(t, float64(42), logLine["count"])
	assert.Equal(t, "acme", logLine["tenant_id"])
	assert.Equal(t, "agent-1", logLine["agent_id"])
	assert.Equal(t, "trace-1", logLine["trace_id"])
}

func TestLogger_LogRawEmptyFields(t *testing.T) {
	var buf bytes.Buffer
	logger := NewLoggerWithWriter("acme", "agent-1", "trace-1", &buf)

	logger.LogRaw(GraphRead, map[string]interface{}{})

	var logLine map[string]interface{}
	err := json.Unmarshal(buf.Bytes(), &logLine)
	require.NoError(t, err)

	assert.Equal(t, "sif.graph_read", logLine["message"])
}

// ---------------------------------------------------------------------------
// Logger — constructor fields
// ---------------------------------------------------------------------------

func TestLogger_FieldsSet(t *testing.T) {
	logger := NewLoggerWithWriter("tenant-x", "agent-y", "trace-z", &bytes.Buffer{})

	assert.Equal(t, "tenant-x", logger.TenantID)
	assert.Equal(t, "agent-y", logger.AgentID)
	assert.Equal(t, "trace-z", logger.TraceID)
}

func TestNewLogger_DefaultsToStdout(t *testing.T) {
	// NewLogger should not panic; it writes to stdout by default.
	logger := NewLogger("tenant-1", "agent-1", "trace-1")
	assert.NotNil(t, logger)
	assert.Equal(t, "tenant-1", logger.TenantID)
}
